File: benchmarks/executable/flat_value_aggregates.c
// Expected: exit 0; stdout: "7,5,26\n"; stderr: ""
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>

typedef struct {
  int64_t first;
  int64_t second;
} TuplePair;

typedef struct {
  int64_t left;
  int64_t right;
} ValuePair;

static TuplePair make_tuple_pair(int64_t first, int64_t second) {
  const TuplePair pair = {first, second};
  return pair;
}

static int64_t combine_tuple_pair(TuplePair pair, int64_t scale) {
  const int64_t product = pair.first * scale;
  return product + pair.second;
}

static ValuePair make_value_pair(int64_t left, int64_t right) {
  const ValuePair pair = {.right = right, .left = left};
  return pair;
}

static int64_t combine_value_pair(ValuePair pair, int64_t scale) {
  const int64_t product = pair.left * scale;
  return product + pair.right;
}

int main(void) {
  const TuplePair tuple = make_tuple_pair(7, 5);
  const ValuePair value = make_value_pair(tuple.first, tuple.second);
  const int64_t tuple_result = combine_tuple_pair(tuple, 3);
  const int64_t value_result = combine_value_pair(value, 3);
  const int64_t result = tuple_result + value_result - 26;
  if (printf("%" PRId64 ",%" PRId64 ",%" PRId64 "\n", value.left,
             value.right, result) < 0) {
    return 1;
  }
  return 0;
}

File: benchmarks/executable/flat_value_aggregates.rs
// Expected: exit 0; stdout: "7,5,26\n"; stderr: ""
#[derive(Clone, Copy)]
struct ValuePair {
    left: i64,
    right: i64,
}

fn make_tuple_pair(left: i64, right: i64) -> (i64, i64) {
    (left, right)
}

fn combine_tuple_pair(pair: (i64, i64), scale: i64) -> i64 {
    let product = pair.0 * scale;
    product + pair.1
}

fn make_value_pair(left: i64, right: i64) -> ValuePair {
    ValuePair { right, left }
}

fn combine_value_pair(pair: ValuePair, scale: i64) -> i64 {
    let product = pair.left * scale;
    product + pair.right
}

fn main() {
    let tuple = make_tuple_pair(7, 5);
    let value = make_value_pair(tuple.0, tuple.1);
    let tuple_result = combine_tuple_pair(tuple, 3);
    let value_result = combine_value_pair(value, 3);
    let result = tuple_result + value_result - 26;
    println!("{},{},{}", value.left, value.right, result);
}
